[WIP] compose: Only show "New Topic" when in stream or topic narrow.

vinitS101 opened the pull request:
"New stream message" is shown instead of "New Topic" when the user is in All messages, Mentions or Starred messages narrows.

Resolves  #12496.

PR is still WIP. More [here](https://chat.zulip.org/#narrow/stream/49-development-help/topic/Referring.20to.20.22home.22.20(.20.2312496)/near/751825)

rishig commented on frontend_tests/node_tests/narrow_activate.js:
```diff
@@ -82,6 +82,7 @@ function test_helper() {
     stub('ui_util', 'change_tab_to');
     stub('unread_ops', 'process_visible');
     stub('compose', 'update_closed_compose_buttons_for_stream');
+    stub('compose', 'update_closed_compose_buttons_for_other_narrows');
```
maybe lets call this update_closed_compose_buttons_for_non_stream_narrows?

vinitS101 replied:
Should it be `update_compose_buttons_for_stream_narrows` or `update_closed_compose_buttons_for_stream_narrows`?

rishig replied:
not sure; we can wait for Tim for that.

vinitS101 replied:
So, the addition of the popover meant that these functions weren't needed anymore. I've removed them entirely.

vinitS101 commented:
I’ll make the change.

rishig commented:
Thanks @vinitS101! 
There must be some code somewhere already that decides whether to prefill the stream (e.g. when you press the `c` shortcut). It would be great to reuse that code.

vinitS101 commented:
@rishig I drastically changed my approach to this issue and made the changes to implement the popover first. I've added a screenshot of what the popover looks like.

vinitS101 commented:
PR is ready for review. @rishig

vinitS101 commented:
Pushed changes with @alexandraciobica 's patch for the placement issue. 
Added screenshots of the popover placement.

@rishig This is ready for review.

vinitS101 commented:
Rebased and made changes to the handlebar templates to follow/work with the new convention that was implemented.

rishig commented on static/js/popovers.js:
```diff
@@ -488,6 +491,59 @@ exports.toggle_actions_popover = function (element, id) {
+        if (filter.has_operator("is") && filter.operands("is")[0] === "private"
+            || filter.has_operator("pm-with") || filter.has_operator("group-pm-with")) {
+            text_stream = i18n.t("New stream message");
+            text_conversation = i18n.t("New conversation");
```
Let's make this "New private message"

vinitS101 replied:
FYI, this change means we don't alter the label of that button at all. So I just moved it to the template.

rishig commented on static/styles/compose.scss:
```diff
@@ -44,11 +44,20 @@
-    padding: 3px 10px;
+    .button.small {
+        font-size: 1em;
+        padding: 3px 10px;
```
I think we'll want the space around 
![image](https://user-images.githubusercontent.com/890911/61175390-6799f880-a563-11e9-8cc5-9db2029c73cb.png)

to be `3px 8px` instead of `3px 10px` (though the rest of the buttons should stay as 10px). 

Also the on-hover of "Reply" is a bit broken -- the right edge doesn't darken on hover. Not a blocker if we can't figure it out, but maybe worth spending a few minutes to see if there is a known solution.

vinitS101 replied:
Made the change from `10px` -> `8px`.

The right border for the reply button was removed just below the quoted CSS here:
```
.compose_reply_button {
        border-radius: 4px 0px 0px 4px;
        border-right: none;
    }
```
I removed it to resemble the reply button from #12334.
On removing `border-right: none;` :

![reply_button_right_border](https://user-images.githubusercontent.com/25329595/61176313-56410280-a5dc-11e9-9968-8e832a89392a.gif)

Should I make this change?

rishig replied:
hm, the double-thickness line definitely isn't good. 
Can the second line be added only on-hover?

vinitS101 replied:
I tried a few variations and it doesn't seem to work on hover.

rishig replied:
hm. Ok, we can deal with that as followup.

rishig commented on static/styles/popovers.scss:
```diff
@@ -354,3 +355,11 @@ ul {
     padding: 2px 0px 3px 0px;
 }
+
+.compose_actions_popover {
```
I tihnk we can just match/reuse the gear menu box-shadow for this?

vinitS101 replied:
The gear menu is the one on the top-right, right? I used the box-shadow CSS that was used for that dropdown-menu.

rishig replied:
Interesting, ok. I remember it being darker, but that could just be because it was on a more-white background. 

If possible it would be great to have it be literally the same CSS (so that changes in one place will get reflected in the other), but if that's complicated then no worries.

vinitS101 replied:
It was the wrong box-shadow that I used. The one that is used for the gear menu is the one `bootstrap.css` has for the `.dropdown-menu` class. But, I used the one for `.nav .dropdown-menu` from `zulip.scss`. I could directly use the `.dropdown-menu` class, but I have made the changes here.

This is what it looks like now:
![Screenshot_5](https://user-images.githubusercontent.com/25329595/61189011-b8fdd100-a6a4-11e9-9552-96eaa24484a5.png)

rishig commented:
Thanks @vinitS101! Left a few minor comments.

One thing I forget if I mentioned is that we should show a "New topic" button when in a stream narrow (in addition to having it in the dropdown).

vinitS101 commented:
I added the 'New Topic' button for streams (hidden in other narrows):

Normal view:
![Screenshot_2](https://user-images.githubusercontent.com/25329595/61188487-4a694500-a69d-11e9-9bad-f1a2f62de7d6.png)

Mobile browser view:
![Screenshot_1](https://user-images.githubusercontent.com/25329595/61188463-155cf280-a69d-11e9-8411-c39470f75035.png)

@rishig Should it be hidden in Mobile browser view?

vinitS101 commented:
I think I've made all the changes that were needed. 
* Added the `New Topic` button for streams.
* Fixed the box-shadow to match the gear menu.
* Changed the button labe from `New conversation` to `New private message` for all narrows.

I was unable to fix the on-hover border issue for the Reply button, but otherwise this PR is ready for the next round of reviews.

vinitS101 commented:
@rishig This is ready for review.

rishig commented on static/js/narrow.js:
```diff
@@ -272,11 +272,14 @@ exports.activate = function (raw_operators, opts) {
-    if (filter.has_operator("is") && filter.operands("is")[0] === "private"
-        || filter.has_operator("pm-with") || filter.has_operator("group-pm-with")) {
-        compose.update_closed_compose_buttons_for_private();
+    if (filter.has_operator("is") && (filter.operands("is")[0] === "private"
```
I still think this is prone to error, and it would be better to have the `if` go in the other direction. I.e. "if in a stream narrow, then .. else ..".

vinitS101 replied:
This was a lot convoluted initially when we had different text for PM narrows and I also kind of followed what we had earlier.
Looking at it now, it can be simplified a lot, and I've made the change. Thanks for bringing this up!

vinitS101 commented:
> The margins for Drafts, Reply etc should still be 10px. The only thing that should be 8px is the down-triangle.

I've made this change. 
![Screenshot_3](https://user-images.githubusercontent.com/25329595/61594260-63c34180-ac07-11e9-9602-560e5b880897.png)

The other two issues seem to be related, although I'm not sure what the reason is. I'll investigate.

vinitS101 commented:
@rishig I pushed a change to the CSS, although I am not sure if that what you intended. Please let me know if this is incorrect.

> If you're in a stream narrow, and click down-triangle -> New topic in .., the cursor focus is set wrong. I haven't read the code around this, but ideally this bug would not be possible, because both the "New topic" button and this would just call the same code.

Also pushed a fix for this issue. 

> If you're in a PM and click "New private message", it should open a PM with no recipients

This bug currently exists in master (tested on czo) and I think it'd be better to address this in a separate issue if that is not a problem.

rishig commented:
> If you're in a stream narrow, and click down-triangle -> New topic in .., the cursor focus is set wrong

hm, I still seem to be running into this, though sometimes it does the right thing.
I think a repro for the broken behavior is: click a topic in the left sidebar, then click down-triangle -> New topic in ...

vinitS101 commented:
You're right. I just tested that and I can reproduce it, although it doesn't happen every time which is pretty confusing.

vinitS101 commented:
Fixed the focus issue and some other smaller changes.

@timabbott this is ready for review.

aero31aero commented on static/js/popovers.js:
```diff
@@ -507,6 +475,51 @@ exports.toggle_actions_popover = function (element, id) {
+    if (!filter) {
+        // All Messagess
+        text_stream = i18n.t("New stream message");
+    } else if (filter.has_operator('stream')) {
```
We could reduce this 3 step if-else to just:

```js
if (filter && filter.has_operator('stream')) {...}
else {...}
```

aero31aero commented on static/js/narrow.js:
```diff
@@ -339,10 +339,12 @@ exports.activate = function (raw_operators, opts) {
-        compose.update_closed_compose_buttons_for_private();
+    if (filter.has_operator('stream')) {
+        // Show the New Topic button in a stream narrows.
+        $(".compose_new_topic_button").attr("hidden", false);
```
Minor nitpick, but having the implementation details of the compose module inside it is cleaner imo, even if it is just two 1 line functions. That way if we rename the button, we just have to change it in `compose.js` and not touch any code that wants to show/hide it.

aero31aero commented on templates/zerver/app/compose.html:
```diff
@@ -15,33 +15,19 @@
-                      id="left_bar_compose_mobile_button_big"
-                      title="{{ _('New message') }} (c)">
-                        <span>+</span>
+                    <button type="button" class="button small rounded compose_new_topic_button"
```
We should use id here instead of class. `$('#compose_new_topic_button')` fits the semantics better.

aero31aero commented on templates/zerver/app/compose.html:
```diff
@@ -15,33 +15,19 @@
                       title="{{ _('Reply') }} (r)">
                         <span class="compose_reply_button_label">{{ _('Reply') }}</span>
+                    </button>{# -#}
+                    <button type="button" class="button small rounded compose_actions_dropdown">
```
Id instead of class here too.

vinitS101 commented:
Thanks for the review, @aero31aero. Made the required changes.

vinitS101 commented:
@timabbott Can you please review this?

vinitS101 commented:
Rebased to remove conflicts.
@timabbott can you please review this?

timabbott commented:
@vinitS101 can you rebase this once more and fix the tests?  We've moved that test suite to puppeteer, so might take a bit of porting but overall puppeteer is simpler :)

timabbott commented:
@vinitS101 it seems to be failing CI; note also that if you rebase, we've ported the Casper tests to Puppeteer (which I expect you'll find a lot nicer to deal with).  See https://zulip.readthedocs.io/en/latest/testing/testing-with-puppeteer.html

timabbott commented:
Closing as https://github.com/zulip/zulip/issues/12496 is no longer the current design plan; thanks for all your work towards this @vinitS101!
